### DISCUSSION TITLE
Why is process.receive not receiving messages?

### DISCUSSION THREAD

**Discussion by Lilja** [gleam-lang/gleam#4083] [Questions & support] [answered]
I'm trying to understand more about subjects and selectors. I made this short snippet, but it's not receiving anything. Why is that? What am I missing?

```gleam
import repeatedly
import gleam/erlang/process

pub fn main() {
  let subject = process.new_subject()
  repeatedly.call(5000, Nil, fn(_state, _i) {
    io.debug("Checking for messages")
    case process.receive(subject, within: 5000) {
      Ok(msg) -> {
        io.debug("Received message " <> string.inspect(msg))
      }
      Error(_) -> {
        io.debug("No message")
      }
    }
    io.debug("Done listening")
    Nil
  })
  repeatedly.call(1000, Nil, fn(_state, _i) {
    io.debug("Sending message")
    process.send(subject, "Hello")
  })
  process.sleep_forever()
}
```

Logs: 

```
"Sending message"
"Sending message"
"Sending message"
"Sending message"
"Checking for messages"
"Sending message"
"Sending message"
"Sending message"
"Sending message"
"Sending message"
"No message"
"Done listening"
"Sending message"
"Sending message"
"Sending message"
"Sending message"
"Sending message"
```

**Comment by lpil** [comment, marked as answer]
The subject does not belong to the process you are attempting to receive using.

You should also ideally not use the repeatedly library like that, instead create a regular process or use the actor module. It's a tool for cross platform scripts and not a building block for OTP programs.

**Reply by Lilja** [reply]
Thanks for answering, Louis!

Does `Selector` have the same design of not allowing a non-process-owner to `process.select` messages? I tweaked the above program to use `process.select` instead of `process.receive` and ran into the same issue. I would think that since I created a selector and selected a subject, it would just forward the messages to the `Selector`.

> You should also ideally not use the repeatedly library like that

This is a simplified version of a grammy(UDP server) and a periodic process that I'm running to check the health of the connected clients. So I'm looking to run something every ten or so seconds. Is that a good use case for using the library? It's always nice to re-use existing software :smiley: 
I suppose I could try writing my own process & actor to contain the state.

**Reply by lpil** [reply]
It's processes themselves that receive messages, not subjects or selectors, they are not channels. Subjects and selectors are references to processes inboxes, they cannot be transferred or shared.

No, that is not a good use of the library. You should be using actors.
